React Queryify SLO

CoenWarmer opened the pull request:
Resolves https://github.com/elastic/kibana/issues/154782

## Summary

This PR integrates React Query in the Observability app, and uses it in the SLO feature.


## Checklist
- [X] Add React Query to o11y app
- [X] Fetch SLOs via RQ hook
- [X] Delete SLO via RQ hook
- [X] Create SLO via RQ hook
- [X] Fetch indices via RQ hook
- [X] Fetch APM Suggestions via RQ hook
- [X] SLO List: Correctly invalidate cache after deleting
- [X] SLO List: Correctly invalidate cache after cloning
- [X] SLO List: Correctly set loading states when deleting or cloning
- [X] Update tests
- [X] Check final types
- [X] Check Storybook
- [X] Optimistic updates when cloning
- [X] Optimistic updates when deleting

CoenWarmer committed:
React Queryify SLO

CoenWarmer committed:
Update tests

kdelemme commented on x-pack/plugins/observability/public/components/app/burn_rate_rule_editor/burn_rate_rule_editor.tsx:
```diff
@@ -13,7 +13,7 @@ import { i18n } from '@kbn/i18n';
 
 import { toDuration, toMinutes } from '../../../utils/slo/duration';
-import { useFetchSloDetails } from '../../../hooks/slo/use_fetch_slo_details';
+import { useFetchSloDetails } from '../../../hooks/slo/use_fetch_slo_details_rq';
```
Are we planning to rename the file without the `_rq` suffix and remove the previous hooks?

CoenWarmer replied:
Yes :) Just updated the PR

kdelemme commented on x-pack/plugins/observability/public/pages/slos/components/slo_list.tsx:
```diff
@@ -94,10 +77,10 @@ export function SloList() {
-          onCloning={handleCloningOrDeleting}
-          onDeleting={handleCloningOrDeleting}
-          onDeleted={handleClonedOrDeleted}
+          onCloned={noop}
+          onCloning={noop}
+          onDeleting={noop}
+          onDeleted={noop}
```
Should we remove these props from the `SloListItems` component if we provide `noop` to them now?

kdelemme commented on x-pack/plugins/observability/public/hooks/slo/use_fetch_slo_details_rq.ts:
```diff
@@ -0,0 +1,55 @@
+
+  const { isInitialLoading, isLoading, isError, isSuccess, isRefetching, data, refetch } = useQuery(
+    {
+      queryKey: ['fetchSloList', sloId],
```
Shouldn't we use `['fetchSloDetails', sloId]` instead?

CoenWarmer replied:
Correct, typo ;)

kdelemme commented on x-pack/plugins/observability/public/hooks/slo/use_delete_slo_rq.ts:
```diff
@@ -0,0 +1,28 @@
+    {
+      mutationKey: ['deleteSlo', sloId],
+      onSuccess: () => {
+        queryClient.invalidateQueries(['fetchSloList']);
```
Should we also include `fetchHistoricalSummary`?

kdelemme commented on x-pack/plugins/observability/public/hooks/slo/use_create_slo_rq.ts:
```diff
@@ -0,0 +1,49 @@
+    {
+      mutationKey: ['updateSlo'],
+      onSuccess: () => {
+        queryClient.invalidateQueries(['fetchSloList']);
```
Should we also include `fetchHistoricalSummary`?

kdelemme reviewed (commented):
Nice work 😍 Really like the simplification React Query brings.. No more useCallback and useMemo required with the useDataFetcher.
Just a couple of questions and suggestions. Also, I haven't try it yet, but do we need to do something for storybook?

shahzad31 commented on x-pack/plugins/observability/public/hooks/slo/use_create_slo_rq.ts:
```diff
@@ -0,0 +1,49 @@
+/*
```
can you please create a new folder in apps with sub folders as sli, slo etc, i think that will simplify code reviews. 

And perhaps create sub directories as components/hooks etc. The root hooks folder etc structure was meant for shared stuff across plugins, apm, uptime, logs etc. 

Also out of curiosity, we have a useFetcher hook in the same folder. wasn't that enough for fetching data?
what are the special feature of react-query that are being utilised?

CoenWarmer replied:
Hey Shazad, everything (components, hooks) are already in separate 'SLO' named folders.

React Query is not a data fetching library so it is not an equivalent to `useFetcher`.

In terms of what React Query can bring:
- less boilerplate (see the diff)
- caching / invalidating per request,
- caching / invalidating requests in different parts of the app so less prop plumbing
- better type safety
- standardization (the data fetching hooks in `o11y` differ quite a bit in terms of return signature)
- built in features such as refetch on window focus, polling, optimistic updates (some of which I use in this PR).

shahzad31 replied:
i wonder why it's triggering the observability-ui reviews then? i think instead of creating sli/slo directory inside other sub directories. You should create an app directory inside apps/actionable-observability/ and that should contain all the relevant components. hooks etc sub directories.

CoenWarmer replied:
Because I've made one change in `x-pack/plugins/observability/public/application/index.tsx` which is owned by `obs-ui`.

shahzad31 replied:
it says i own 14 files in the PR, not just index file :)

shahzad31 replied:
<img width="1758" alt="image" src="https://user-images.githubusercontent.com/3505601/220301128-ae7e7920-f4d9-4ccb-9bf8-641493ab574b.png">
 

hooks/slo codeowner needs to be updated. this is why i am requesting to create a new app directory here 
like 
`https://github.com/elastic/kibana/tree/main/x-pack/plugins/observability/public/apps/actionable_observability`

and keep everything related inside. Updating each new dir code-owner doesn't seems ideal.

CoenWarmer replied:
Or we could just update the CODEOWNERS file right now and be done with it. I'll do that now.

CoenWarmer replied:
Updated. Let me know if I missed anything that might still trigger your notifications.

CoenWarmer replied:
I assume your concerns have been addressed by https://github.com/elastic/kibana/pull/151672. Resolving this.

CoenWarmer committed:
Assorted fixes

CoenWarmer commented:
> Nice work 😍 Really like the simplification React Query brings.. No more useCallback and useMemo required with the useDataFetcher. Just a couple of questions and suggestions. Also, I haven't try it yet, but do we need to do something for storybook?

Yes! Storybook still needs some love. Will look into it tomorrow.

CoenWarmer committed:
Fix Storybook

CoenWarmer committed:
Fix tests

CoenWarmer committed:
Merge branch 'main' of github.com:elastic/kibana into feature/react-query-in-o11y

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

CoenWarmer committed:
Fix types

CoenWarmer committed:
Merge branch 'feature/react-query-in-o11y' of github.com:CoenWarmer/kibana into feature/react-query-in-o11y

CoenWarmer committed:
Update CODEOWNERS file

CoenWarmer committed:
Fix types

CoenWarmer committed:
Fix tests

CoenWarmer committed:
Merge branch 'main' of github.com:elastic/kibana into feature/react-query-in-o11y

CoenWarmer committed:
Revert CODEOWNERS

CoenWarmer committed:
Merge branch 'main' of github.com:elastic/kibana into feature/react-query-in-o11y

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/generate codeowners'

CoenWarmer committed:
Update CODEOWNERS

CoenWarmer committed:
Merge branch 'feature/react-query-in-o11y' of github.com:CoenWarmer/kibana into feature/react-query-in-o11y

CoenWarmer committed:
Update CODEOWNERS

kdelemme commented on x-pack/plugins/observability/public/hooks/slo/use_create_slo.ts:
```diff
@@ -5,75 +5,91 @@
-      setLoading(true);
-      setError('');
-      setSuccess(false);
+  const cloneSlo = useMutation<
```
🍰 I wonder if this hook is not doing to many things: create, clone and update. I think it would make sense to split it into 3 hooks: use_clone_slo, use_create_slo, use_update_slo

CoenWarmer replied:
Yeah, that might be wise, as it would mean these mutating hooks will have the same return signature as the delete slo hook. Going to do that!

CoenWarmer commented on x-pack/plugins/observability/public/hooks/slo/use_create_slo.ts:
```diff
@@ -5,75 +5,91 @@
-    createSlo,
-    updateSlo,
-  };
+  return { createSlo, cloneSlo, updateSlo };
```
All these objects have their own `isLoading`, `isError`, `isSuccess`, `error`, `mutate`, `mutateAsync`, etc. inside them so you have more granular control about which request succeeded or failed.

kdelemme commented on x-pack/plugins/observability/public/hooks/slo/use_delete_slo.ts:
```diff
@@ -5,42 +5,57 @@
+        // Snapshot the previous value
+        const previousSloList = queryClient.getQueryData<FindSLOResponse>(['fetchSloList']);
+
+        // Optimistically update to the new value
```
👍🏻  Very nice to have this optimistic handling.

kdelemme commented on x-pack/plugins/observability/public/hooks/slo/use_fetch_apm_suggestions.ts:
```diff
@@ -29,48 +37,36 @@ interface ApiResponse {
-
-  return { data: suggestions, loading, error };
-}
+  const { isInitialLoading, isLoading, isError, isSuccess, isRefetching, data, refetch } = useQuery(
```
👍🏻 The code is so much simpler to reason about.

CoenWarmer commented on x-pack/plugins/observability/public/hooks/slo/use_create_slo.ts:
```diff
@@ -5,75 +5,91 @@
+      return http.post<CreateSLOResponse>(`/api/observability/slos`, { body });
+    },
+    {
+      onMutate: async ({ slo, idToCopyFrom }) => {
```
This is the optimistic update to immediately show a cloned result without waiting for the API to come back with the result. If we would that the `slo` value here and use it to do `queryClient.setQueryData`, it would be the summary version, leading to the app to break. So we cheat a little, and take the id to copy the existing SLO that's inside the query cache, and use that.

kdelemme commented on x-pack/plugins/observability/public/pages/slo_edit/components/slo_edit_form.tsx:
```diff
@@ -67,44 +66,56 @@ export function SloEditForm({ slo }: Props) {
-      const processedValues = transformValuesToUpdateSLOInput(values);
-      updateSlo(slo.id, processedValues);
+      try {
+        const processedValues = transformValuesToUpdateSLOInput(values);
+
+        await updateSlo.mutateAsync({ sloId: slo.id, slo: processedValues });
+
+        toasts.addSuccess(
+          i18n.translate('xpack.observability.slos.sloEdit.update.success', {
+            defaultMessage: 'Successfully updated {name}',
+            values: { name: getValues().name },
+          })
+        );
```
👍🏻  Pfew, that's nice. No more useEffect elsewhere to handle the success/error cases.

kdelemme commented on x-pack/plugins/observability/public/utils/kibana_react.storybook_decorator.tsx:
```diff
@@ -5,14 +5,45 @@
       services={{
         application: {
           navigateToUrl: () => {},
+          capabilities: {
```
👍🏻

kdelemme reviewed (commented):
Second review, just a nit that we can tackle in another PR. Going to play with it locally before approving.
Good job 👏🏻

kdelemme commented:
When testing locally, the first call to the historical summary api returns super fast without data, and then it takes a couple dozen of seconds to fetch again the data.

![image](https://user-images.githubusercontent.com/1376800/220407266-89268395-1f22-4d90-b630-cd495ff306bf.png)

CoenWarmer commented:
Thanks for these comments @kdelemme! Code still needs a bit of refinement; will pick this up soon.

CoenWarmer committed:
Break up useCreateSlo hook into three separate hooks

CoenWarmer committed:
Fix fetching historical summary

CoenWarmer committed:
Disable fetching slo detail api when no sloId is passed

CoenWarmer committed:
Fix fetching of fetchSloList params

CoenWarmer committed:
Update components which used one useCreateOrUpdateSlo hook

CoenWarmer commented:
> Going to update this comment as I found things:
> 
> 1. When loading the page, the first call to the historical summary api returns super fast without data, and then it takes a couple dozen of seconds to fetch again the data.

I can't reproduce this anymore; I think it might be the result of `refetchOnWindowFocus` being set to `true`. More on that below.

> ![image](https://user-images.githubusercontent.com/1376800/220407266-89268395-1f22-4d90-b630-cd495ff306bf.png)

> 2. If you have the consol open, click in the consol and then in the page (focus, unfocus the page basically), API calls a re triggered.

This is a RQ feature that's on by default, but we can turn off: https://tanstack.com/query/v4/docs/react/guides/window-focus-refetching

I agree we should turn it off for now, as it has the weird side effect that it's no longer possible to select an element using the element inspector in Chrome. 

> 3. Searching, Filtering and Sorting does not work

Fixed.

> 4. When the Create SLO page loads, an API call to `http://localhost:5601/kibana/api/observability/slos/undefined` is made.

Fixed.

> 5. When you create an SLO and are redirected to the dashboard page, the charts are not populated.

Fixed.

> Recording of the main issues I have locally: [user-images.githubusercontent.com/1376800/220409422-9998f4cb-0a35-4f63-9c72-2d25bed8f6c1.mov](https://user-images.githubusercontent.com/1376800/220409422-9998f4cb-0a35-4f63-9c72-2d25bed8f6c1.mov)

CoenWarmer committed:
Fix type errors

CoenWarmer committed:
Reintroduce optimistic UI updates for cloning and deleting

CoenWarmer committed:
Merge branch 'main' of github.com:elastic/kibana into feature/react-query-in-o11y

CoenWarmer committed:
Make queryKey for useFetchHistoricalSummary deterministic

CoenWarmer committed:
Remove useState/useEffect when getting sloIds to pass to useFetchHistoricalSummary hook

kdelemme reviewed (approved):
Thanks for addressing the issues and splitting the hooks in 3.

One last storybook issue for the slo burn rate rule:
![image](https://user-images.githubusercontent.com/1376800/220767642-7faa1b9c-696b-4b23-a2d2-b076ab929a49.png)

Approving as it works well locally 👍🏻 Good job
